corehq/apps/userreports/es/data_source.py

```diff
@@ -0,0 +1,234 @@
+
+
+class ConfigurableReportEsDataSource(ReportDataSource):
+    def __init__(self, domain, config_or_config_id, filters, aggregation_columns, columns, order_by):
+        self.lang = None
+        self.domain = domain
+        if isinstance(config_or_config_id, DataSourceConfiguration):
+            self._config = config_or_config_id
+            self._config_id = self._config._id
+        else:
+            assert isinstance(config_or_config_id, basestring)
+            self._config = None
+            self._config_id = config_or_config_id
+
+        self._filters = {f.slug: f for f in filters}
+        self._filter_values = {}
+        self._deferred_filters = {}
+        self._order_by = order_by
+        self._aggregation_columns = aggregation_columns
+        self._column_configs = OrderedDict()
+        for column in columns:
+            # should be caught in validation prior to reaching this
+            assert column.column_id not in self._column_configs, \
+                'Report {} in domain {} has more than one {} column defined!'.format(
+                    self._config_id, self.domain, column.column_id,
+                )
+            self._column_configs[column.column_id] = column
+
+    @property
+    def aggregation_columns(self):
+        # TODO add deferred filters to support mobile ucr
+        return self._aggregation_columns
+
+    @property
+    def config(self):
+        if self._config is None:
+            self._config, _ = get_datasource_config(self._config_id, self.domain)
+        return self._config
+
+    @property
+    def column_configs(self):
+        return self._column_configs.values()
+
+    @property
+    def table_name(self):
+        # TODO make this the same function as the adapter
+        return get_table_name(self.domain, self.config.table_id).lower()
+
+    @property
+    def filters(self):
+        return filter(None, [f.to_es_filter() for f in self._filter_values.values()])
+
+    def set_filter_values(self, filter_values):
+        for filter_slug, value in filter_values.items():
+            self._filter_values[filter_slug] = self._filters[filter_slug].create_filter_value(value)
+
+    def defer_filters(self, filter_slugs):
+        self._deferred_filters.update({
+            filter_slug: self._filters[filter_slug] for filter_slug in filter_slugs})
+
+    def set_order_by(self, columns):
+        self._order_by = columns
+
+    @property
+    def order_by(self):
+        if self._order_by:
+            return [
+                (col.field, order)
+                for sort_column_id, order in self._order_by
+                for col in [self._column_configs[sort_column_id]]
+            ]
+        elif self.column_configs:
+            col = self.column_configs[0]
+            return [(col.field, ASCENDING)]
+        return []
+
+    @property
+    def columns(self):
+        db_columns = [c for sql_conf in self.es_column_configs for c in sql_conf.columns]
```

Comment:
not a big deal but might want to change `sql_conf` to `es_conf` or something like that